### PR TITLE
Settings objects and example uses

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Introducing the classes that can eliminate widely separated duplication of literals for the path and default value of a preference.

Instead global variables of these types will hold all the information in one place, and users of the variables are forced to have compile and link dependency on them -- a good thing to be explicit about at build time, avoiding the brittleness of the coincident literals.

There is also caching of preference values in memory.

Some example uses.  The large, routine task of converting many other uses of gPrefs remains.

These examples also invert or break a few source file dependencies in a good way.

### CODE REVIEW

File: src/Prefs.h [hunk 1]
```diff
@@ -55,6 +57,168 @@ extern int gMenusDirty;
...
+{
+public:
+   using Setting::Setting;
+
+   //! Write the negation of the previous value, and then return the current value.
+   bool Toggle();
+};
+
+//! Specialization of Setting for int
+class IntSetting final : public Setting< int >
```
Comment (crsib) [inline src/Prefs.h:202]: Not that it really matters, but feels like `using IntSetting = Setting<int>;` looks cleaner in this context

File: src/AudioIOBase.cpp [hunk 2]
```diff
@@ -197,7 +197,8 @@ void AudioIOBase::HandleDeviceChange()
       captureParameters.suggestedLatency =
          Pa_GetDeviceInfo(recDeviceNum)->defaultLowInputLatency;
    else
-      captureParameters.suggestedLatency = DEFAULT_LATENCY_CORRECTION/1000.0;
+      captureParameters.suggestedLatency =
+         AudioIOLatencyCorrection.GetDefault()/1000.0;
```
Comment (crsib) [inline src/AudioIOBase.cpp:201]: We really should think about (gradually) switching to std::chrono at some point.

**Review by crsib — APPROVED** [review]
There are few minor comments, that could be addressed.

Do we plan to extract settings into a library later on? It could be useful for all kinds of modules, so they don't have to link against Audacity directly.

File: src/Prefs.h [hunk 3]
```diff
@@ -55,6 +57,168 @@ extern int gMenusDirty;
 struct ByColumns_t{};
 extern ByColumns_t ByColumns;
 
+//! Base class for settings objects.  It holds a configuration key path.
+/* The constructors are non-explicit for convenience */
+class AUDACITY_DLL_API SettingBase
+{
+public:
+   SettingBase( const char *path ) : mPath{ path } {}
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply src/Prefs.h:65]: Maybe std::wstring.

std::string is little used in our code.  wxString predominates.  That's of course a tolerable wxBase only dependency.

File: src/Prefs.h (hunk 1 above)
Comment (Paul-Licameli) [reply src/Prefs.h:202]: I needed to derive BoolSetting to add a method.  Doing this for the other specializations just looked more consistent.  Not very important.

File: src/AudioIOBase.cpp (hunk 2 above)
Comment (Paul-Licameli) [reply src/AudioIOBase.cpp:201]: Yes.  Maybe than a Settings<std::chrono::milliseconds> ?

File: src/Prefs.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (crsib) [reply src/Prefs.h:65]: Well, it makes sense to be moving to UTF-8 gradually. std::wstring has different behavior on Windows and Unix

**Comment by Paul-Licameli** [comment]
> There are few minor comments, that could be addressed.
> 
> Do we plan to extract settings into a library later on? It could be useful for all kinds of modules, so they don't have to link against Audacity directly.

If you study my big graph of intended libraries, you will see lib-preferences is one and very low-level.  It will be one of the first taken out.

File: src/Prefs.h (hunk 1 above)
Comment (Paul-Licameli) [reply src/Prefs.h:202]: Another advantage of doing it as I did, is that it's easier to forward-declare IntSetting etc. as incomplete types.

**Pull request merged**